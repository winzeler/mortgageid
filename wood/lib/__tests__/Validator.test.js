const { Validator } = require('#lib/Validator');
const { Rule } = require('#lib/Rules');
const { ERROR_EMPTY } = require('#lib/Errors');

describe('Validator', () => {
  describe('errors', () => {
    let validator;

    beforeEach(() => {
      validator = new Validator(['abc', 'def']);
      validator.rules = {
        abc: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'abc is empty.' }),
        ],
        def: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'def is empty.' }),
        ],
      };
    });

    // ---------------------------------------------------------------------------------------------

    it('should get failing rules for multiple fields', () => {
      const errors = validator.errors({
        abc: '',
        def: '',
      }, {});

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should get failing rule for single field', () => {
      const errors = validator.errors({
        abc: '',
        def: '',
      }, {}, 'def');

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should incorporate api errors for multiple fields', () => {
      const errors = validator.errors({
        abc: '',
        def: 'not empty',
      }, {
        abc: [{
          code: 'ERROR_API',
          title: 'An API error for abc.',
        }],
        def: [{
          code: 'ERROR_API',
          title: 'An API error for def.',
        }],
      });

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should incorporate api errors for single fields with existing error', () => {
      const errors = validator.errors({
        abc: '',
        def: 'not empty',
      }, {
        abc: [{
          code: 'ERROR_API',
          title: 'An API error for abc.',
        }],
        def: [{
          code: 'ERROR_API',
          title: 'An API error for def.',
        }],
      }, 'abc');

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should incorporate api errors for single fields with no existing error', () => {
      const errors = validator.errors({
        abc: '',
        def: 'not empty',
      }, {
        abc: [{
          code: 'ERROR_API',
          title: 'An API error for abc.',
        }],
        def: [{
          code: 'ERROR_API',
          title: 'An API error for def.',
        }],
      }, 'def');

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should return empty when no errors', () => {
      const errors = validator.errors({
        abc: 'abc',
        def: 'def',
      }, {
        abc: [],
        def: [],
      });

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should get failing rules for aliased fields', () => {
      validator = new Validator(['abc', 'def'], { aliases: { an_alias: 'def' } });
      validator.rules = {
        abc: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'abc is empty.' }),
        ],
        def: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'def is empty.' }),
        ],
      };

      const errors = validator.errors({
        abc: 'abc',
        an_alias: '',
      }, {
        abc: [],
        an_alias: [],
      });

      expect(errors).toMatchSnapshot();
    });

    // ---------------------------------------------------------------------------------------------

    it('should return empty when aliased fields have no errors', () => {
      validator = new Validator(['abc', 'def'], { aliases: { an_alias: 'def' } });
      validator.rules = {
        abc: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'abc is empty.' }),
        ],
        def: [
          new Rule('isEmpty', { code: ERROR_EMPTY, title: 'def is empty.' }),
        ],
      };

      const errors = validator.errors({
        abc: 'abc',
        an_alias: 'def',
      }, {
        abc: [],
        an_alias: [],
      });

      expect(errors).toMatchSnapshot();
    });
  });
});
