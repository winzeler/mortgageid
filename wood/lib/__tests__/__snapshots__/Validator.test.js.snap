// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator errors should get failing rule for single field 1`] = `
Array [
  Object {
    "code": "ERROR_EMPTY",
    "title": "def is empty.",
  },
]
`;

exports[`Validator errors should get failing rules for aliased fields 1`] = `
Object {
  "an_alias": Array [
    Object {
      "code": "ERROR_EMPTY",
      "title": "def is empty.",
    },
  ],
}
`;

exports[`Validator errors should get failing rules for multiple fields 1`] = `
Object {
  "abc": Array [
    Object {
      "code": "ERROR_EMPTY",
      "title": "abc is empty.",
    },
  ],
  "def": Array [
    Object {
      "code": "ERROR_EMPTY",
      "title": "def is empty.",
    },
  ],
}
`;

exports[`Validator errors should incorporate api errors for multiple fields 1`] = `
Object {
  "abc": Array [
    Object {
      "code": "ERROR_EMPTY",
      "title": "abc is empty.",
    },
    Object {
      "code": "ERROR_API",
      "title": "An API error for abc.",
    },
  ],
  "def": Array [
    Object {
      "code": "ERROR_API",
      "title": "An API error for def.",
    },
  ],
}
`;

exports[`Validator errors should incorporate api errors for single fields with existing error 1`] = `
Array [
  Object {
    "code": "ERROR_EMPTY",
    "title": "abc is empty.",
  },
  Object {
    "code": "ERROR_API",
    "title": "An API error for abc.",
  },
]
`;

exports[`Validator errors should incorporate api errors for single fields with no existing error 1`] = `
Array [
  Object {
    "code": "ERROR_API",
    "title": "An API error for def.",
  },
]
`;

exports[`Validator errors should return empty when aliased fields have no errors 1`] = `Object {}`;

exports[`Validator errors should return empty when no errors 1`] = `Object {}`;
